### PR TITLE
Implement SpECTRE singleton as Charm++ single-element array

### PR DESCRIPTION
**Pull request by knelli2**
## Proposed changes

This is a rebased and cleaned up version of #3354. The bulk is largely the same though.

Instead of having the spectre singleton be based on a charm++ singleton chare, this PR implements the spectre singleton as a single element charm array chare. This is for a few reasons:

1. Charm++ singletons cannot be (easily) placed on particular processors. Typically we will want singletons on their own processors.
2. Charm++ singletons don't participate in load balancing.
3. Charm++ singletons don't participate in checkpoint/restart when restarted on a different number of procs.

To make this change easier, this PR first wraps calls to `proxy.ckLocal()` and `proxy.ckLocalBranch()` in new wrapper functions `Parallel::local(proxy)` and `Parallel::local_branch(proxy)`.

Since this is an involved change, there are a lot of commits detailing each change. After reviews are done, I can squash some together if people want.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Parallel/Invoke.hpp
Comment (nilsvu) [inline src/Parallel/Invoke.hpp:72]: Move the next commit that adds `Parallel::local_branch` before this one.

File: src/Parallel/Local.hpp
Comment (nilsvu) [inline src/Parallel/Local.hpp:46]: Could you add a bit of documentation here and/or to `Parallel::local` that describes their difference, and what they are used for? Would it make sense to use `Parallel::local` for both, with an `if constexpr` switch? I have no idea what's the difference between `ckLocal()` and `ckLocalBranch()`.

File: src/Parallel/Local.hpp
Comment (nilsvu) [inline src/Parallel/Local.hpp:35]: Is there a way to assert the array indeed has only a single element here?

File: src/Parallel/Main.hpp
Comment (nilsvu) [inline src/Parallel/Main.hpp]: (nitpick) typo: their

Comment (nilsvu) [inline src/Parallel/Main.hpp]: What are your plans for distributing singletons to cores? Do you think we should allow each singleton component to choose the core by itself, or perhaps have the metavars of every executable decide which set of cores is available for each parallel component? I'm thinking we may have a number of cheap singletons that are absolutely fine sharing PE 0 with DG elements, and blocking a PE for each singleton may not be the best. For example, the elliptic solver uses a bunch of "residual monitor" singletons that are very cheap and don't need to run on their own private core. Also, I expect interpolation targets (which are singleton components) often don't need to run on their own private core, except of course of expensive things like the AH finder.

**Review by nilsvu — CHANGES_REQUESTED** [review]
Looks good overall, couple of comments.

**Comment by nilsvu** [comment]
I tested this on Minerva, and at least the elliptic solver (which uses two singletons) runs fine on two nodes. The two singletons are placed on PEs 0 and 1 (both on node 1).

File: src/Parallel/Local.hpp
Comment (knelli2) [reply src/Parallel/Local.hpp:35]: I don't think so. The docs [here](https://charm.readthedocs.io/en/latest/faq/manual.html#is-there-a-property-similar-to-thisindex-containing-the-chare-arrays-dimensions-or-total-number-of-elements) say that because array chares can have their size changed dynamically, that `any such value would potentially be stale between access and use`.

I can add an `is_singleton<Proxy>` check that checks if the chare type is `Singleton`.

File: src/Parallel/Local.hpp
Comment (knelli2) [reply src/Parallel/Local.hpp:46]: Could you elaborate on what is confusing/missing from the current documentation? They are wrappers around `.ckLocal()` and `.ckLocalBranch()` which return pointers to their local objects.

From reading the charm docs, I don't know *why* they are different but charm makes it a point to only use `.ckLocal()` on arrays/array-elements and `.ckLocalBranch()` on groups/nodegroups. For that reason, I think it's best to keep them separate.

File: src/Parallel/Main.hpp
Comment (knelli2) [reply src/Parallel/Main.hpp]: I'm currently trying to come up with a scheme to best distribute everything. However, this will involve modifying where we put DG elements because they can no longer be on all cores if we are reserving some. So this will be a separate PR. For now I settled to just put each singleton on its own core where DG elements can also occupy those cores.

I agree that probably not all singletons need their own cores. The control system singletons themselves are also fairly cheap so most won't need their own core (except maybe size control). I'll take this into account. Maybe the singletons themselves have a `constexpr bool` saying they need their own core?

File: src/Parallel/Local.hpp
Comment (nilsvu) [reply src/Parallel/Local.hpp:46]: Well if from the user's (our) perspective `ckLocal()` and `ckLocalBranch()` are the same, only named differently for the chare types, then I think it would be nice if we didn't have to care about it. `Parallel::local` can take care of calling the right function depending on the type.

File: src/Parallel/Main.hpp
Comment (nilsvu) [reply src/Parallel/Main.hpp]: Yea let's decide these things in a follow-up PR, the round-robin distribution you chose here for now is good.

When coming up with a scheme, please take into account that we'll probably need to experiment with it, so don't make everything compile-time choices. Once things are implemented at compile-time then it's hard to switch back to runtime choices. Generally, prefer input-file options where a compile-time choice doesn't give any benefits (this is my view).

File: src/Parallel/Local.hpp
Comment (kidder) [reply src/Parallel/Local.hpp:46]: As stated in the documentation, `ckLocalBranch` is always guaranteed to return a pointer to an existing object, either the element of the nodegroup on the node (or the element of the group on the PE) on which the call is made.  

`ckLocal` can only be called on a singleton or element of an array, returning either a pointer to the local object if it exists on the local processor and returns `NULL` if the element does not exist or is on a different processor

Therefore technically you should check to see if a ckLocal call didn't return NULL before using the pointer, but in most (if not all) of our use cases we are getting an object from its own proxy

File: src/Parallel/Invoke.hpp
Comment (kidder) [reply src/Parallel/Invoke.hpp:72]: actually just move this to the commit that makes the change from `ckLocalBranch` to `Parallel::local_branch`

File: src/Parallel/Local.hpp
Comment (kidder) [reply src/Parallel/Local.hpp:35]: I don't think we want to change this function to accept an `array_proxy`; instead (if possible) it would be better to make sure that what is passed into this function is always an `array_element_proxy`

File: src/Parallel/Algorithms/AlgorithmSingleton.hpp
Comment (kidder) [inline src/Parallel/Algorithms/AlgorithmSingleton.hpp:44]: what is the error if this include is not added?  would it be sufficient to replace `AlgorithmSingleton.def.h` with `AlgorithmArray.def.h`?

File: src/Parallel/Algorithm.hpp
Comment (kidder) [inline src/Parallel/Algorithm.hpp]: used to uniquely identify

and remove spatially

File: src/Parallel/Main.hpp
Comment (kidder) [inline src/Parallel/Main.hpp]: build -> built

**Review by kidder — CHANGES_REQUESTED** [review]

File: src/Parallel/Local.hpp
Comment (knelli2) [reply src/Parallel/Local.hpp:38]: In the docs for these functions, it says that it is the responsibility of the user (not this function) to check that the pointer is valid. I think this makes it ok to combine the functions into a single `Parallel::local`

File: src/Parallel/Local.hpp
Comment (knelli2) [reply src/Parallel/Local.hpp:35]: I like @kidder's suggestion. This actually will simplify things a lot.

File: src/Parallel/Local.hpp
Comment (nilsvu) [reply src/Parallel/Local.hpp:46]: Thanks for explaining @kidder. I think we should combine both into a single `Parallel::local` then. It is guaranteed to return a non-null pointer for nodegroups and groups (since there's always a local object), and it may return null for array elements (since they may live on a different core). Sound good?

Comment (knelli2) [reply src/Parallel/Local.hpp:46]: I understand where you're coming from, however, Charm expressly named the functions differently for a reason. Whatever that reason may be, I think we should just mimic them and not combine things ourselves.

That being said, @kidder if you think I should combine them, then I will.

Comment (nilsvu) [reply src/Parallel/Local.hpp:46]: I think you're putting too much trust into decisions made by Charm++ here :D

File: src/Parallel/Algorithms/AlgorithmSingleton.hpp
Comment (knelli2) [reply src/Parallel/Algorithms/AlgorithmSingleton.hpp:44]: Changing it to `AlgorithmArray.def.h` gives a bunch of redefinition errors. Removing the `AlgorithmArray.hpp` include gives a bunch of `implicit instantiation of undefined template` errors all over the place. Seems like it's necessary.

File: src/Parallel/Local.hpp
Comment (knelli2) [reply src/Parallel/Local.hpp:35]: I've thought about this a lot and I actually now disagree with @kidder. This function is typically used as:
```
auto& proxy = Parallel::get_parallel_component<ParallelComponent>(cache);
...*Parallel::local(proxy)...;
```

For array elements, it makes sense you want to pass in a proxy to the element like `*Parallel::local(proxy[array_index])`, because you're not calling things on the full array, you're calling them on the elements.

However, for singletons, the user doesn't (and shouldn't) care how the singleton is implemented. So calling this on a singleton proxy should just be `*Parallel::local(proxy)`, not `*Parallel::local(proxy[0])`. The latter puts the burden on the user to remember how singletons are implemented (as single element chare-arrays with an index `int` that is 0) which I don't think we should do.

Thoughts?

File: src/Parallel/Local.hpp
Comment (fmahebert) [reply src/Parallel/Local.hpp:46]: @nilsvu The reason I opted to mirror the Charm++ design when originally putting this together was exactly what @kidder said above: `ckLocal` and `ckLocalBranch` have different behavior because they act on different objects. In most cases, I felt the programmer _should_ be aware of what chare is being used in a piece of code, and how that impacts the validity of the obtained local pointer. I felt that maintaining two distinct names for two (similar but) different functions was reasonable. Anyway, that was my thinking at the time, but feel free to change however you like 🙂

File: src/Parallel/Local.hpp
Comment (fmahebert) [reply src/Parallel/Local.hpp:35]: @knelli2 Yes, you've identified why I went with the current implementation: to hide the details of the singleton implementation inside `local` so that the user doesn't have to worry about it.

But it's true that adding some check that the Proxy is intended to be a singleton would help avoid a potential bug where someone passes in a (non-singleton) array but forgets to index it. I've been away from this for too long to remember if that's easy to do or not 😅

**Comment by knelli2** [comment]
Added some fixups and a new commit. Will fix the "wrong commit" comment when I squash.

File: src/Parallel/TypeTraits.hpp
Comment (nilsvu) [inline src/Parallel/TypeTraits.hpp]: Just use `std::is_same` directly. Shallow wrappers like this often make code harder to understand (they hide simple code under a new name, they require you to remember they exist and what they do, and they need to be maintained and tested).

File: src/Parallel/TypeTraits.hpp
Comment (nilsvu) [inline src/Parallel/TypeTraits.hpp]: same comment about using `std::is_same` directly. You can make `get_parallel_component_from_proxy` public.

File: tests/Unit/Parallel/Test_TypeTraits.cpp
Comment (nilsvu) [inline tests/Unit/Parallel/Test_TypeTraits.cpp]: These "Failed testing X" messages are a relict from the past. You don't need them when the `static_assert` is descriptive enough by itself.

**Review by nilsvu — CHANGES_REQUESTED** [review]
you can edit the fixups with my comments directly to make things easier for @kidder.

File: src/Parallel/Local.hpp
Comment (nilsvu) [reply src/Parallel/Local.hpp:46]: > In most cases, I felt the programmer should be aware of what chare is being used in a piece of code, and how that impacts the validity of the obtained local pointer.

Definitely 👍

> I felt that maintaining two distinct names for two (similar but) different functions was reasonable.

Here I disagree. The two names have no difference in meaning AFAICT; they might as well be called `ckLocal()` and `ckLocal2()` :D The programmer should remember which chare they are working with, and hence, if they need to check the result for `nullptr`, but they should not have to remember which of the two function names they need. Even if the names are different they have to remember the pointer validity, so different function names don't help there.

Either way, I don't feel strongly enough about this to hold up the PR. If @kidder agrees I suggest @knelli2 you merge `Parallel::local` and `Parallel::local_branch`, and if @kidder disagrees then don't.

**Comment by knelli2** [comment]
Squashed in @nilsvu's changes into the fixups.

File: src/Parallel/Local.hpp
Comment (knelli2) [reply src/Parallel/Local.hpp:46]: @kidder ping

**Comment by knelli2** [comment]
Rebased on develop, fixed conflict, and updated `ControlSystem/WriteData.hpp` to use `Parallel::local()` (was a new file). No other changes.

File: src/Parallel/Local.hpp
Comment (knelli2) [reply src/Parallel/Local.hpp:46]: @kidder ping

Comment (kidder) [reply src/Parallel/Local.hpp:46]: I would like to suggest that `Parallel::local` be renamed to `Parallel::if_local_get_element` and `Parallel::local_branch` be renamed to `Parallel::get_local_element_of_group`

Comment (nilsvu) [reply src/Parallel/Local.hpp:46]: @kidder I see your wish to choose more precise names. I think we shouldn't put "element" in the names, since that can be confused with our notion of domain elements. Also, one of the functions works with a singleton, which is just a single entity. My proposal is: in this case of controversy let's go with minimal changes here and keep the Charm++ names `local` and `local_branch` (or if you agree with my comment above @kidder we merge them under the name of `local`).

**Comment by kidder** [comment]
okay, given the desire of getting this in, please rebase and squash keeping local and local_branch as separate names.

File: src/Parallel/Local.hpp
Comment (nilsdeppe) [reply src/Parallel/Local.hpp:46]: I think we should match the Charm++ names, at the very least to not diverge in naming from them too much. If we start renaming everything then we can't defer to their documentation as much, and I think deferring to the Charm++ documentation makes our life easier overall. I think this is the argument I made last time this came up for discussion and why Francois went with this distinction.

**Comment by nilsdeppe** [comment]
Please ping me once it's rebased and squashed and I'll look through it!

**Comment by knelli2** [comment]
@nilsdeppe @kidder @nilsvu Squashed and rebased! :)

File: src/Parallel/Invoke.hpp
Comment (nilsdeppe) [reply src/Parallel/Invoke.hpp:72]: This is still in the wrong commit

**Review by nilsdeppe — COMMENTED** [review]
LGTM, just one ping from a previous review by others. Please squash immediately

File: src/Parallel/Invoke.hpp
Comment (knelli2) [reply src/Parallel/Invoke.hpp:72]: Ah yes I missed this. Thank you

**Review by nilsvu — DISMISSED** [review]

**Review by nilsdeppe — DISMISSED** [review]

**Comment by kidder** [comment]
@nilsvu please investigate mac CI failure

**Comment by kidder** [comment]
@knelli2 @nilsvu the failing test also fails running charm v7 on my linux desktop, so it does not appear to be a mac specific issue

**Comment by kidder** [comment]
also, the test passes on develop with charm v7

**Comment by knelli2** [comment]
@kidder I am able to reproduce the failure on charm v7. However if I run `ctest -R Unit.Parallel.AlgorithmParallelWithBalancing --repeat until-pass:1000`, the test will pass after 5-10 tries. Now sure why yet.

**Comment by kidder** [comment]
I find this test executable to be opaque with several magic numbers that are not documented.   It looks like it was hijacked for testing load balancing and checkpointing.   I am thinking of writing a simpler, well-documented one more suitable for debugging when the test fails.  I should have it done by Monday.

**Comment by knelli2** [comment]
Rebased and fixed conflicts.

@kidder I agree. I had a very hard time trying to understand what was going on in the test (honestly I still don't 😅)

**Comment by knelli2** [comment]
Now macOS passed? Very confusing.

**Comment by nilsdeppe** [comment]
@kidder it sounded like this is a Charm++ 7 related issue with load balancing. How should we proceed? This change is also needed for the memory monitor that @knelli2 is adding and using to debug the BBH issues, so it would be nice to at least have an update on what the next steps for this PR are :)

**Comment by kidder** [comment]
If possible, don't run the load-balancing test when using charm7.  Then this can be merged

**Comment by nilsdeppe** [comment]
From the spectre core dev meeting the conclusion was to disable the load balancing test with Charm++ 7 for now:

```
   if (CHARM_VERSION VERSION_EQUAL 7.0.0)
     message(STATUS "Periodic load balancing with the RotateLB is not working \
 correctly with Charm++ 7.0.0. Disabling the test for now until this is \
 resolved.")
   else()
     add_algorithm_test_with_balancing("AlgorithmParallelWithBalancing"
       "AlgorithmParallel" "num_objs=14")
   endif()
```
in `tests/Unit/Parallel/CMakeLists.txt` line 150-ish. Can you do that in a separate commit? I'm worried that not each commit in this PR will build and pass tests. Specifically, there's a commit `Update parallel LB test` implying that some commits before this will fail tests. This needs to be resolved (and tested to make sure each commit builds and passes tests) because otherwise doing something like `git bisect` will be very difficult

**Comment by knelli2** [comment]
@nilsdeppe @nilsvu @kidder
- Added a commit that disables the LB test for charm 7
- Combined multiple related commits into a single commit to address @nilsdeppe's concerns about each commit building and passing independent of each other (detailed all changes in commit message). I check all commits where major changes occurred and confirmed they all build and pass on their own. (I didn't bother checking the commits that just had documentation changes)
- Also tested with charm 7 and all tests passed on my linux machine.
- No other code changes

**Review by nilsdeppe — APPROVED** [review]

**Comment by nilsdeppe** [comment]
@kidder @nilsvu can you please look at this again? This is currently a major blocker on the BBH front for merging everything back into develop...

**Review by nilsvu — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Comment by fmahebert** [comment]
Thanks for seeing this through @knelli2 🎉

**Comment by knelli2** [comment]
Thank you @fmahebert for laying the groundwork! 👏🏻

**Pull request merged**
